src/browser/BrowserService.cpp

```diff
@@ -629,11 +629,12 @@ QJsonObject BrowserService::showPasskeysRegisterPrompt(const QJsonObject& public
     const auto excludeCredentials = publicKey["excludeCredentials"].toArray();
     const auto attestation = publicKey["attestation"].toString();
 
-    // Check Resident Key requirement
+    // Non-Resident Keys are not supported
     const auto authenticatorSelection = publicKey["authenticatorSelection"].toObject();
     const auto requireResidentKey = authenticatorSelection["requireResidentKey"].toBool();
-    if (requireResidentKey) {
-        return getPasskeyError(ERROR_PASSKEYS_RESIDENT_KEYS_NOT_SUPPORTED);
+    const auto residentKey = authenticatorSelection["residentKey"].toString();
+    if (residentKey == BrowserPasskeys::REQUIREMENT_DISCOURAGED || (residentKey.isEmpty() && !requireResidentKey)) {
```

Comment:
You don't need to error if the client doesn't require a resident key, an authenticator can choose to create a credential as a resident key anyway, see https://www.w3.org/TR/2019/REC-webauthn-1-20190304/#op-make-cred, steps 4 and 7.4.

Reply:
Agreed

Reply:
This should only return if:
- `residentKey` is set to `discouraged`
- `residentKey` has not been set, and `requireResidentKey` is missing or set to `false`

Is this against those steps? I'm not sure what I'm missing here.

Reply:
An authenticator is always allowed to create resident keys (it doesn't matter whether they're required, preferred or discouraged by the RP), so this logic makes KeePassXC unnecessarily incompatible with RPs that make requests that meet those conditions.

Reply:
The spec is weirdly silent or at least unclear on that front. I do agree there is no _prohibition_ against always creating a resident key.

Reply:
It's not easy to follow, but it is explicitly allowed. In https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#sctn-createCredential:

![image](https://github.com/keepassxreboot/keepassxc/assets/2430738/bcb28c6e-f282-4694-a0bf-abe49843a842)

Then later in https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#sctn-op-make-cred: 

![image](https://github.com/keepassxreboot/keepassxc/assets/2430738/e5998bf8-f51e-4551-a4ba-abba2b24fa6d)

Specifically that "or the authenticator chooses".

Reply:
Can there be a situation where `residentKey` is `discouraged` and the server does not accept Resident Keys?

Reply:
Not according to the spec, see the note box here: https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#dom-residentkeyrequirement-discouraged

The RP is never explicitly told if a credential is a resident key or not. It can guess based on the size of the credential ID, since non-resident keys can be implemented by having the credential ID be the private key encrypted by a key stored in the authenticator, but an authenticator could choose to provide long credential IDs for resident keys.